### ISSUE TITLE
Volume mount license into Windows Containers

### ISSUE THREAD

**Issue by samcook** [InRule/Samples#13]
Hi,

I've attempted to make use of volume mounts to inject the InRuleLicense.xml file into a Windows Container image at runtime (which would mean you don't need to make a specific image for any given machine you want to run it on).

`docker run -d --restart unless-stopped --name ircatalog -e CatalogUser=xxx -e CatalogPassword=yyy -e CatalogDbHost=zzz -e CatalogDbName=ircatalog-db -p 8100:80 -v "C:\Users\Sam\Documents\InRule\License\:C:\ProgramData\InRule\SharedLicenses\:ro" server/inrule-catalog:5.0.26`

Unfortunately, I haven't had much luck at actually getting InRule to pick up the license using this method. I've tried mounting a folder containing the license file in the same location as one that's baked into an image, but it always says it's running in grace period mode when using the volume mounted license folder.

The only difference I can see, inside the container, is that the volume mounted folder is a symlink directory, instead of a "real" folder. I'm able to read the license file inside the symlinked folder though.

`11/21/2017  02:39 PM    <SYMLINKD>     TestVolume [\\?\ContainerMappedDirectories\71953C40-5FB5-44FC-ADDF-43763C99372E]`

Any ideas why this might be the case?

**Comment by jelster** [comment]
Hi @samcook thanks for raising this. 

From your docker run line, I am seeing that you are looking to bind-mount the license folder as a volume, but it looks like the path may not be correct on the container's side of things; the license file location should have spaces in it - `C:\Program Data\InRule\Shared Licenses\`. 

I've found the spaces don't mesh well with the traditional volume syntax in Windows, so I switched to the `--mount` syntax, which the Docker docs recommend anyway. 
  
https://docs.docker.com/engine/reference/commandline/run/#mount-volume--v-read-only

Something like this I think is how it ends up:
`docker run --mount type=bind,src='C:\Program Data\InRule\SharedLicenses\'...`

An alternative to mounting the volume into the Program Data directory is to mount it into the `/bin` directory of the service. On the container (except for the WebCatMan image), this is exposed as the environment variable `ContainerDir` (webcatman: `c:\inrule-catalog-manager`).

If what you're seeing though is that the volume is mounting correctly but the service isn't picking up the license, then it may need to have the AppPool recycled/restarted -- try bumping it with an iisreset

Let me know how this goes!

**Comment by samcook** [comment]
Hi @jelster 

Thanks for the quick reply.

First up, are you sure the path should have spaces? The license file location on my local Windows 10 machine is `C:\ProgramData\InRule\SharedLicenses`, and I was getting a log message in the catalog container that it was running in grace mode when I connected the Catalog Manager until I changed the license file location (see pull request #12, the previous Dockerfile appeared to be copying the license to a file called `C:\Program Data\InRule\Shared Licenses`, rather than into a directory of that name).

I've just tried using the mount parameter syntax, but it appears to do the same thing as the volume syntax (when I inspect the container, either way it says the volume is a bind mount).

Would mounting it into the `bin` directory work? With Windows containers you can only mount folders, not individual files, so unless it knew to look in a subfolder I don't think that would work?

I'll have another go giving it an iisreset and report back.

**Comment by samcook** [comment]
Hi @jelster 

I've done a bit more testing.

In order to rule out issues with the ProgramData path, spaces, etc, I added the following to the catalog's `web.config` file:

```
  <inrule.repository>
    <licensing licenseFolder="C:\\InRuleLicense" />
  </inrule.repository>
```

When I create an image with a the license file baked into it at that path, I don't get any license messages in the log when I connect Catalog Manager to the running container.

When I create an image with no baked in license file, and use a bind mount at that location, I get a the following message output in the log:
```
Time: 11/21/2017 7:44:28.146 PM, Source: Licensing, Message: 
               The irCatalog® grace period license will expire 11/26/2017. Use 
               the Activation Utility to manage your InRule product 
               activations.
               Installer Version: 5.0.26.120
               irSDK Version: 5.0.26.120
```

The command used to start the container was:
`docker run -d --restart unless-stopped --name ircatalog -e CatalogUser=xxx -e CatalogPassword=yyy -e CatalogDbHost=zzz -e CatalogDbName=ircatalog-db -p 8100:80 --mount type=bind,src="C:\Users\SuperUser\Documents\InRule\License",dst="C:\InRuleLicense",ro server/inrule-catalog:5.0.26`

Here's how they look from within the container (first one is baked in license file directory, second is bind mount)
```
11/21/2017  07:43 PM    <DIR>          InRuleLicense
11/21/2017  07:43 PM    <SYMLINKD>     InRuleLicense2 [\\?\ContainerMappedDirectories\B34A9CE7-9BA2-47E0-90E7-5FA24DCECD08]
```

Both files are readable and identical
```
C:\>fc c:\InRuleLicense\InRuleLicense.xml c:\InRuleLicense2\InRuleLicense.xml
Comparing files C:\INRULELICENSE\InRuleLicense.xml and C:\INRULELICENSE2\INRULELICENSE.XML
FC: no differences encountered
```

Performing an `iisreset` didn't change anything - still got log entries about grace period licensing.

Is there something in the license checking code that doesn't like symlinks?

**Comment by jelster** [comment]
Hmm you're right about the lack of spaces in the path -- I don't know where I went astray on that :)

Thanks for the troubleshooting info -- I'll take a look and ask around to see if anyone knows anything

**Comment by samcook** [comment]
Great, let me know if I can be of any further assistance.

Would be awesome if there were official InRule docker images, just add license and away you go 👍

**Comment by samcook** [comment]
Hi @jelster 

I was thinking more on this overnight, and wondered if the issue was due to the IIS application pool identity within the container not having permission to read the bind mounted license file (even though the "root" user in the container could).

I created a simple test web application that echoed back the contents of a specified file, build an image with that, and ran some tests.

When I mounted the license file from my home directory (i.e. `C:\Users\Sam\Documents\InRule\License`) the web app wasn't able to read the file (though the "root" container user could).

When I mounted the license file from a directory with more generous permissions (i.e. `C:\InRuleLicense`) the web app read the file just fine.

Problem solved!

**Comment by jelster** [comment]
that's great news! In retrospect, I should've remembered this sooner - I think avoiding the permissions issue is one of the reasons I went with the baked-in license approach for my initial PR. 

So if I'm understanding the workaround correctly, the solution is to use a directory *other* than the `C:\ProgramData\InRule\SharedLicenses`, mounting the volume as discussed above in all other respects, right?

Thanks again for helping out with this and sleuthing!

**Comment by samcook** [comment]
It's fine to use `C:\ProgramData\InRule\SharedLicenses` as the target inside the container - the source just needs to have permissions such that the application pool identity can read it (which wasn't the case when I was originally mapping from a subdirectory of my home directory `C:\Users\Sam\...`)

I now have the catalog service running in a docker container with the license file mounted from `C:\InRule\License` on the host to `C:\ProgramData\InRule\SharedLicenses` in the container.

The only change I ended up making to the Dockerfile was to remove the line that copies in the baked-in license.

**Comment by jelster** [comment]
Hmmm now that you mention it, why not just replace the license file copy line like you did and replace it with a RUN line that sets perms for the folder (a la how the runtime and catalog grant the app pool identity perms to the container dir):

**Comment by samcook** [comment]
The problems that I can see there are:
- the directory doesn't exist at image build time, as we are mounting it into the container at run time
- if you create the directory beforehand (and docker only seems to lets you mount a volume to an existing directory if the target directory is empty), I suspect that the volume mount "hides" the empty directory with the symlink, so I don't know that it would make a difference
- if you try and change the permission at runtime (e.g. in the start script or something) it may not work (e.g. the volume is mounted read-only)

IMO the permission issue is really an "issue" on the docker host, not in the guest container - the same thing could happen for any container if the process running inside the container doesn't have permission to access the file on the outside.

It's up to the person running the container to make sure that they set up the permissions correctly (I just happened to stuff it up the first time I tried it!). Maybe add a note in the readme for running the container?
